### PR TITLE
Add unit tests for RegisterStreamer

### PR DESCRIPTION
**Pull request by YangchenYe323**
### What changes are proposed in this pull request?
Created a unit test class for `RegisterStreamer` to test expected interaction and stream operations.

### Why are the changes needed?
- Improve unit test coverage(0% ->92%)
- Exposed some under-clarified behavior in the class

### Does this PR introduce any user facing changes?
No

### CODE REVIEW

File: core/server/worker/src/test/java/alluxio/worker/block/RegisterStreamerTest.java [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+
+  /**
+   * Create a mocked RegisterStreamer Instance for test use.
+   *
+   * @param mockChannel mock channel for grpc redirection
+   * @param workerId mock worker id for this streamer
+   * @param numBlocks number of blocks in the stream, used to control concurrent requests
+   * @return mocked RegisterStreamer instance
+   */
+  private RegisterStreamer createMockRegisterStreamer(
```
Comment (dbw9580) [inline core/server/worker/src/test/java/alluxio/worker/block/RegisterStreamerTest.java:360]: this can be static

File: core/server/worker/src/test/java/alluxio/worker/block/RegisterStreamerTest.java [hunk 2]
```diff
@@ -0,0 +1,402 @@
...
+              }
+
+              @Override
+              public void onError(Throwable t) {
+              }
+
+              @Override
+              public void onCompleted() {
+              }
+            };
```
Comment (dbw9580) [inline core/server/worker/src/test/java/alluxio/worker/block/RegisterStreamerTest.java:99]: You can create an base class `NoOpStreamObserver` with empty impls for each method, and then extends this class in each test, overriding the methods that actually need to do something.

Comment (dbw9580) [inline core/server/worker/src/test/java/alluxio/worker/block/RegisterStreamerTest.java:99]: To save some boilerplate, you can create a base class `NoOpStreamObserver` with empty impls for each method, and then extends this class in each test, overriding the methods that actually need to do something.

File: core/server/worker/src/test/java/alluxio/worker/block/RegisterStreamerTest.java [hunk 3]
```diff
@@ -0,0 +1,402 @@
...
+          }
+        });
+
+    // the streamer should throw DeadlineExceededException
+    // because the request gets no response
+    RegisterStreamer streamer = createMockRegisterStreamer(mChannel, 0L, 1);
```
Comment (dbw9580) [inline core/server/worker/src/test/java/alluxio/worker/block/RegisterStreamerTest.java:135]: Why workerId == 0?

Comment (dbw9580) [inline core/server/worker/src/test/java/alluxio/worker/block/RegisterStreamerTest.java:135]: Why workerId == 0?

File: core/server/worker/src/test/java/alluxio/worker/block/RegisterStreamerTest.java [hunk 4]
```diff
@@ -0,0 +1,402 @@
...
+
+    RegisterStreamer streamer = createMockRegisterStreamer(mChannel, 0L, 2);
+    assertThrows(InternalException.class, streamer::registerWithMaster);
+  }
+
+  @Test(timeout = 5000)
+  public void testRegisterSuccess() throws Exception {
+    // store workerId to requestCount
+    ConcurrentMap<Long, AtomicInteger> requestCount = new ConcurrentHashMap<>();
+    // create a server that finally works
```
Comment (dbw9580) [inline core/server/worker/src/test/java/alluxio/worker/block/RegisterStreamerTest.java:274]:
```suggestion
    // create a server that accepts and completes the registration properly
```

Comment (dbw9580) [inline core/server/worker/src/test/java/alluxio/worker/block/RegisterStreamerTest.java:274]:
```suggestion
    // create a server that finally works
```
```suggestion
    // create a server that accepts and completes the registration properly
```

File: core/server/worker/src/test/java/alluxio/worker/block/RegisterStreamerTest.java [hunk 5]
```diff
@@ -0,0 +1,402 @@
...
+  private GrpcChannel mChannel;
+
+  @Test(timeout = 5000)
+  public void testRegisterTimeOutConcurrentRequest() throws Exception {
+    // create a server that does not respond to requests
+    createMockService(
+        new BlockMasterWorkerServiceGrpc.BlockMasterWorkerServiceImplBase() {
+          @Override
+          public StreamObserver<RegisterWorkerPRequest> registerWorkerStream(
+              StreamObserver<RegisterWorkerPResponse> responseObserver) {
```
Comment (dbw9580) [inline core/server/worker/src/test/java/alluxio/worker/block/RegisterStreamerTest.java:86]: Can create a `WorkerRegistrationServiceHandler` class that is subclass of `BlockMasterWorkerServiceImplBase` and takes 3 handler methods that are `onNext` etc., and make `createMockService` take an instance of this class.

Comment (dbw9580) [inline core/server/worker/src/test/java/alluxio/worker/block/RegisterStreamerTest.java:86]: Can create a `WorkerRegistrationServiceHandler` class that is subclass of `BlockMasterWorkerServiceImplBase` and takes 3 handler methods that are `onNext` etc., and make `createMockService` take an instance of this class.

**Review by dbw9580 — CHANGES_REQUESTED** [review]
First batch of comments

File: core/server/worker/src/test/java/alluxio/worker/block/RegisterStreamerTest.java (hunk 3 above)
Comment (YangchenYe323) [reply core/server/worker/src/test/java/alluxio/worker/block/RegisterStreamerTest.java:135]: It shouldn't be. The parameter exists so the test methods can verify the streamer sends the correct workerId. It should be nonzero so as not to be confused with empty protobuf value.

File: core/server/worker/src/test/java/alluxio/worker/block/RegisterStreamerTest.java (hunk 5 above)
Comment (YangchenYe323) [reply core/server/worker/src/test/java/alluxio/worker/block/RegisterStreamerTest.java:86]: I didn't quite understand the suggestion.. Are you suggesting creating a class like the following?
```
class WorkerRegistrationServiceHandler extends BlockMasterWorkerServiceImplBase {
  WorkerRegistrationServiceHandler(function onNext, function onError, function onCompleted) {
  }
}
```

Comment (YangchenYe323) [reply core/server/worker/src/test/java/alluxio/worker/block/RegisterStreamerTest.java:86]: I thought about this again, there might be some danger in the encapsulation we're trying to make.
The below class:
```java
class WorkerRegistrationServiceHandler 
    extends BlockMasterWorkerServiceImplBase {

    StreamObserver mObserver;

    WorkerRegistrationServiceHandler(StreamObserver observer) {
        mObserver = observer;
    }

    @Override
      public StreamObserver<RegisterWorkerPRequest> registerWorkerStream(
          StreamObserver<RegisterWorkerPResponse> responseObserver) {
        //! BIG DANGER: mObserver will re re-used for a second call
        return mObserver;
      }
}
```
is NOT correct in the strict sense of a GRPC handler, as it reused possibly closed streams for further requests. The test functions in fact make one requests, but this code could easily be a problem if someone re-used it for new test functions.

The most decent solution would be to take, not an observer instance, but a function to create an instance. But this requires something like Factory pattern in java(:<), which could really result in more boilerplate than we're trying to eliminate. 

For now I would probably suggest staying with the boilerplate with the improvement of `NoOpStreamObserver`

Comment (YangchenYe323) [reply core/server/worker/src/test/java/alluxio/worker/block/RegisterStreamerTest.java:86]: An alternative is to make a check inside `registerWorkerStream` method to only handle the first request. But this is still awkward and can easily create problems for new tests.

File: core/server/worker/src/main/java/alluxio/worker/block/RegisterStreamer.java [hunk 6]
```diff
@@ -221,13 +221,13 @@ private void registerInternal() throws InterruptedException, DeadlineExceededExc
...
-    if (mFinishLatch.getCount() == 0) {
-      abort();
-    }
-
     // Wait for all batches have been ACK-ed by the master before completing the client side
     if (!mAckLatch.await(mResponseTimeoutMs * MAX_BATCHES_IN_FLIGHT, TimeUnit.MILLISECONDS)) {
+      // If the master side is closed before the client side, there is a problem
+      if (mFinishLatch.getCount() == 0) {
+        abort();
+      }
```
Comment (dbw9580) [inline core/server/worker/src/main/java/alluxio/worker/block/RegisterStreamer.java:229]: @YangchenYe323 can you leave some comments on why this change is necessary?
@jiacheliu3 can you take a look here?

File: core/server/worker/src/test/java/alluxio/worker/block/RegisterStreamerTest.java (hunk 5 above)
Comment (YangchenYe323) [reply core/server/worker/src/test/java/alluxio/worker/block/RegisterStreamerTest.java:86]: Yeah that's a good idea, but `StreamObserver` need to access a `responseObserver` to propogate response to the other side. There's still the problem how this observer can be passed into the `NoOpStreamObserver`.
```java
private static class TestRegistrationServiceHandler
      extends BlockMasterWorkerServiceGrpc.BlockMasterWorkerServiceImplBase {
    private Supplier<StreamObserver<RegisterWorkerPRequest>> mSupplier;

    private TestRegistrationServiceHandler(
        Supplier<StreamObserver<RegisterWorkerPRequest>> supplier) {
      mSupplier = supplier;
    }

    @Override
    public StreamObserver<RegisterWorkerPRequest> registerWorkerStream(
        StreamObserver<RegisterWorkerPResponse> responseObserver) {
      // how to pass the responseObserver to the stream?
      return mSupplier.get();
    }
  }
```

File: core/server/worker/src/main/java/alluxio/worker/block/RegisterStreamer.java (hunk 6 above)
Comment (YangchenYe323) [reply core/server/worker/src/main/java/alluxio/worker/block/RegisterStreamer.java:229]: Sure. The `mFinishLatch` check is moved after the `mAckLatch` check to catch more `InternalException` and `CancelledExeption` cases, not let them be masked as a `DeadlineExceededException`. Before the code looks like:
```java
if (mFinishLatch.getCount() == 0) {
        abort();
}
if (!mAckLatch.await(mResponseTimeoutMs * MAX_BATCHES_IN_FLIGHT, TimeUnit.MILLISECONDS)) {
...
throw DeadlineExceededException
```
It is easy for the server to raise an error **after** the finish check because not network latency, and this error will fail to be captured by the client.

Comment (jiacheliu3) [reply core/server/worker/src/main/java/alluxio/worker/block/RegisterStreamer.java:229]: Makes sense, it is better to do the finish check after some waiting. Good catch!

In your example,
```
send all Worker requests
if (mFinishLatch.getCount() == 0) {
        abort();
}
if (!mAckLatch.await(mResponseTimeoutMs * MAX_BATCHES_IN_FLIGHT, TimeUnit.MILLISECONDS)) {
...
throw DeadlineExceededException
```
After all worker requests are sent, the worker should wait before expecting the responses from master, because the master needs time to process those requests.

File: core/server/worker/src/test/java/alluxio/worker/block/RegisterStreamerTest.java [hunk 7]
```diff
@@ -0,0 +1,364 @@
...
+  private static final InetSocketAddress TEST_ADDRESS = new InetSocketAddress("localhost", 9999);
+
+  @Rule
+  public ConfigurationRule mConfigurationRule =
+      new ConfigurationRule(
+          new ImmutableMap.Builder<PropertyKey, Object>()
+          // set response timeout short to ensure that tests don't take too long
+          .put(PropertyKey.WORKER_REGISTER_STREAM_RESPONSE_TIMEOUT, 500)
+          .put(PropertyKey.WORKER_REGISTER_STREAM_DEADLINE, 2000)
+          .put(PropertyKey.WORKER_REGISTER_STREAM_COMPLETE_TIMEOUT, 500)
```
Comment (jiacheliu3) [inline core/server/worker/src/test/java/alluxio/worker/block/RegisterStreamerTest.java:67]: Forgot if these properties accept strings. If they do, strings like "500ms" or "2s" are more expressive because they come with units.

File: core/server/worker/src/test/java/alluxio/worker/block/RegisterStreamerTest.java (hunk 5 above)
Comment (jiacheliu3) [reply core/server/worker/src/test/java/alluxio/worker/block/RegisterStreamerTest.java:86]: I believe Bowen's suggestion (and the consensus from our engineers) is to NOT use mocks. The reason is, if you have a class Foo, and you `mock(Foo.class)`, then later Foo implementation is changed, the change will be silently mocked. Then in a test against mocked Foo, our assumptions may not hold true anymore. An extended child class, on contrast will complain if there's an API change, so your tests fail and you know where to look at.

@YangchenYe323 On your concern, generally we don't worry about that too much. A child class used merely as a mock does not need to be logically complete. If the implementation makes it hard to be reused, the easiest way out is to add a comment saying do not reuse in other tests because of blah.

File: core/server/worker/src/test/java/alluxio/worker/block/RegisterStreamerTest.java [hunk 8]
```diff
@@ -0,0 +1,364 @@
...
+          // we don't handle authentication in testing
+          .put(PropertyKey.SECURITY_AUTHENTICATION_TYPE, AuthType.NOSASL)
+          .build(),
+          Configuration.modifiableGlobal());
+
+  private GrpcServer mServer;
+  private GrpcChannel mChannel;
+
+  @Test(timeout = 5000)
+  public void testRegisterTimeOutConcurrentRequest() throws Exception {
```
Comment (jiacheliu3) [inline core/server/worker/src/test/java/alluxio/worker/block/RegisterStreamerTest.java:80]: nit
```suggestion
  public void testRegisterTimeoutConcurrentRequests() throws Exception {
```

File: core/server/worker/src/test/java/alluxio/worker/block/RegisterStreamerTest.java [hunk 9]
```diff
@@ -0,0 +1,364 @@
...
+    );
+
+    mServer = GrpcServerBuilder.forAddress(
+        GrpcServerAddress.create(TEST_ADDRESS),
+        Configuration.global(),
+        ServerUserState.global()
+    )
+        .addService(ServiceType.BLOCK_MASTER_WORKER_SERVICE, new GrpcService(mockService))
+        .build()
+        .start();
```
Comment (jiacheliu3) [inline core/server/worker/src/test/java/alluxio/worker/block/RegisterStreamerTest.java:291]: nit: I think our coding style is more like
```suggestion
    mServer = GrpcServerBuilder.forAddress(
          GrpcServerAddress.create(TEST_ADDRESS),
          Configuration.global(),
          ServerUserState.global())
        .addService(ServiceType.BLOCK_MASTER_WORKER_SERVICE, new GrpcService(mockService))
        .build()
        .start();
```

**Review by jiacheliu3 — CHANGES_REQUESTED** [review]
Added some comments and suggestions. Nice work!

File: core/server/worker/src/test/java/alluxio/worker/block/RegisterStreamerTest.java (hunk 5 above)
Comment (YangchenYe323) [reply core/server/worker/src/test/java/alluxio/worker/block/RegisterStreamerTest.java]: I've made a fix in the new commit. Actually mocking is redundant in this test class, because we have been creating a real `BlockMasterWorkerServiceImplBase` anyway.

File: core/server/worker/src/test/java/alluxio/worker/block/RegisterStreamerTest.java (hunk 7 above)
Comment (YangchenYe323) [reply core/server/worker/src/test/java/alluxio/worker/block/RegisterStreamerTest.java]: Fixed

File: core/server/worker/src/test/java/alluxio/worker/block/RegisterStreamerTest.java (hunk 8 above)
Comment (YangchenYe323) [reply core/server/worker/src/test/java/alluxio/worker/block/RegisterStreamerTest.java]: Fixed

File: core/server/worker/src/test/java/alluxio/worker/block/RegisterStreamerTest.java (hunk 9 above)
Comment (YangchenYe323) [reply core/server/worker/src/test/java/alluxio/worker/block/RegisterStreamerTest.java]: Fixed

File: core/server/worker/src/test/java/alluxio/worker/block/RegisterStreamerTest.java [hunk 10]
```diff
@@ -77,16 +76,9 @@ public class RegisterStreamerTest {
   private GrpcChannel mChannel;
 
   @Test(timeout = 5000)
-  public void testRegisterTimeOutConcurrentRequest() throws Exception {
+  public void testRegisterTimeOutConcurrentRequests() throws Exception {
```
Comment (jiacheliu3) [inline core/server/worker/src/test/java/alluxio/worker/block/RegisterStreamerTest.java:79]: nit: `registerTimeoutConcurrentRequests`

Comment (jiacheliu3) [reply core/server/worker/src/test/java/alluxio/worker/block/RegisterStreamerTest.java:79]: and this applies to all other test methods

**Review by jiacheliu3 — APPROVED** [review]
LGTM with some minor comments, good to go once those comments are resolved. Thanks!

File: core/server/worker/src/test/java/alluxio/worker/block/RegisterStreamerTest.java (hunk 10 above)
Comment (YangchenYe323) [reply core/server/worker/src/test/java/alluxio/worker/block/RegisterStreamerTest.java]: Fixed in new commit
